### DISCUSSION TITLE
How i can get name uploaded files during record creation

### DISCUSSION THREAD

**Discussion by pgswow** [pocketbase/pocketbase#1171] [Q&A]
Hello! How i can get name uploaded files during record creation? I need save original file name, if file name is not the Latin alphabet, but for example Chinese characters or Cyrillic, a new filename is generated.

I think I can update the collection after uploading, but I can't find out the original filename.

**Comment by ganigeorgiev** [comment]
Currently there is no easy way to change the uploaded file name without "hacky" workarounds.

You can get the original file name by reading the MultipartForm data after record creation, but you'll also have to rename the uploaded file and this currently can be done only if you use the local filesystem (the filesystem abstraction doesn't support renaming file keys; I'll consider updating it in the future). Here is a short example (I haven't tested it!):

```go
app.OnRecordAfterCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	_, fh, _ := e.HttpContext.Request().FormFile("yourFileField")

	customFileName := fh.Filename // eventually do some normalizations...

	oldFileName := e.Record.GetString("yourFileField")

	e.Record.Set("yourFileField", customFileName)
	if err := app.Dao().SaveRecord(e.Record); err != nil {
		return err
	}

	// Rename the uploaded file in the local storage dir
	//
	// NB! There are no security checks for path traversal!
	oldFilePath := filepath.Join(app.DataDir(), "storage", e.Record.BaseFilesPath(), oldFileName)
	newFilePath := filepath.Join(app.DataDir(), "storage", e.Record.BaseFilesPath(), customFileName)
	os.Rename(oldFilePath, newFilePath)

	return nil
})
```

It depends on your use case, but in general I'd recommend NOT allowing custom upload filenames, because you also will have to take care for sanitizing the user input and making sure that the final path format is supported on the targeted platform (eg. has less than 256 chars; for a list with all filesystem limitations you could check https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits).

If you want to just show a human readable uploaded filename to the end user, you can define a custom text field (eg. `filename`) to your collection and set it to the original name during upload client-side, or auto assign it server-side similar to the above but using the before hooks:

```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	_, fh, _ := e.HttpContext.Request().FormFile("yourFileField")

	// auto store the original uploaded filename (if not explicitly set)
	//
	// we don't need to explicitly call save because the finalizer
	// of the before event hook will do it for us
	if e.Record.GetString("filename") != "" {
		e.Record.Set("filename", fh.Filename)
	}

	return nil
})
```

**Reply by pgswow** [reply]
I do not need to change the file name on the server, what you have now implemented works fine and does not need to be changed, I meant exactly the second option you described.

I just want to add a filename field and write the original or human-readable file name there, but what if there are more than one file? I need to match each uploaded file to a filename. {_5jkAHw0INc_7gUipJoj8L.pdf: myOriginalFile.pdf }

**Reply by ganigeorgiev** [reply]
@pgswow 

> what if there are more than one file?

For multiple uploaded files, you can move the files in their own collection (eg. `files`), where each file will be a single record and link them back in the main collection with a `relation` type field (this could be useful also if you want to add additional meta data to a single file, eg. alt text, description, tags, etc.).

Another workaround would be to transliterate/normalize the file names client-side before upload.

**Reply by ganigeorgiev** [reply]
_In any case, I'll consider in the future (most likely together with the private/protected files) to store the original filename as metadata so that you'll be able to retrieve it by doing something like:_

```go
fs, _ := app.NewFilesystem()
defer fs.Close()

attrs, _ := fs.Attributes(fikekey)
attrs.Metadata["originalName"]
```

**Reply by pgswow** [reply]
Thanks! =)

**Reply by ganigeorgiev** [reply]
@pgswow Just to let you know that the original uploaded name was added to the metadata in `master` and it will be available with the next release. The metadata key is `original_filename` (_gocloud.dev/blob always normalizes all metadata keys to lower case_) and it will contain the first 255 chars of the filename (the length restriction is to prevent large `.attrs` files).

**Reply by calumk** [reply]
@ganigeorgiev 

> > what if there are more than one file?
> 
> For multiple uploaded files, you can move the files in their own collection (eg. `files`), where each file will be a single record and link them back in the main collection with a `relation` type field (this could be useful also if you want to add additional meta data to a single file, eg. alt text, description, tags, etc.).
> 

This seems like a slightly problematic way of doing things, especially if building on a running system where "multiple files" are already uploaded using the "multiple files" options, it would be hard to migrate.

I think ideally at a minimum, it should be possible to display a text "Name" with each file, and perhapse consider how files are returned to the user.

At the moment, we get back just the name - it would be better to recieve something like:

```js
{
    filename : "myfile_fobcaiucsbi.jpeg",
    originalname : "myfile.jpeg",
    mime : "image/jpeg",
    name : "My File",
    filesize : 100,
}
```
If you wanted to do it without breaking existing API, it could maybe be implemented by using "expand" on the files field?

At the moment the lack of this info means i have to display files to users as follows:

<img width="400" alt="image" src="https://github.com/pocketbase/pocketbase/assets/1183991/33902bdc-2eb9-4ee6-9535-e62042e19db2">

instead of :

<img width="200" alt="image" src="https://github.com/pocketbase/pocketbase/assets/1183991/448dfed5-93d6-4d6f-bcb6-a1d46a8a2393">

**Reply by ganigeorgiev** [reply]
@calumk I'm not sure that we can use `expand` for files because it will complicate too much the internals as the `expand` is reserved for relational Record models, but I'll at least consider for the next release exposing an API endpoint for returning the uploaded file attributes.

**Reply by calumk** [reply]
Hi @ganigeorgiev , Any further thoughts on if this is on the roadmap?

**Reply by ganigeorgiev** [reply]
I still haven't got the time to think on it. It exists in the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1), but most likely it will not be part of the upcoming v0.17.0 release as other things got prioritized. There are no ETAs for this feature.

**Reply by mikekok** [reply]
Any updates on this?
I'm working on a project where custom file naming is a must and I really want to use pocketbase (like I do with most of my projects recently).

**Reply by ganigeorgiev** [reply]
@mikekok I'm not sure to what actually you are referring to but nothing has changed since my last comment and there are no plans for now to allow custom uploaded file names out of the box.
In the future there will be a route that will return the meta attributes of the file like size, mimetype and the original filename but this remains a low priority for now.

**Reply by mikekok** [reply]
> @mikekok I'm not sure to what actually you are referring to but nothing has changed since my last comment and there are no plans for now to allow custom uploaded file names out of the box. In the future there will be a route that will return the meta attributes of the file like size, mimetype and the original filename but this remains a low priority for now.

I was referring to changing the filename. Ok, I understand. Thank you for letting me know.
